sql: add basic support for procedure privileges

mgartner opened the pull request:
#### sql/parser: support PROCEDURE in privilege-related statements

This commit adds parsing support for `PROCEDURE` keywords within
privilege-related statements include `GRANT`, `REVOKE`, and
`ALTER DEFAULT PRIVILEGES`.

Release note: None

#### sql: rename function privileges to routine privileges

This commit renames "function" privileges to "routine" privileges
because these privileges are also used for procedures, and "routine" is
the more general term that includes both functions and procedures.

Also, the `tree.FuncObj` type  has been renamed to `tree.RoutineObj` to
be more general. Related types, variables, and functions have be
similarly renamed.

Release note: None

#### sql: add basic support for procedure privileges

This commit implements the basic logic for granting, revoking, and
authorizing privileges on procedures. These operations perform the same
as privilege-related operations on functions, so the tests in
`procedure_privileges` have largely been copied from `udf_privileges`.
There's a number of statements that are not yet supported, so many of
the copied tests are commented out for now. They will be addressed in a
future commit.

Epic: CRDB-25388

Release note: None

#### upgrades: skip TestGrantExecuteToPublicOnAllFunctions under race

TestGrantExecuteToPublicOnAllFunctions is very slow under the `--race`
test flag, often timing-out in CI. This commit skips this test under the
`--race` flag.

Release note: None

mgartner committed:
sql/parser: support PROCEDURE in privilege-related statements

This commit adds parsing support for `PROCEDURE` keywords within
privilege-related statements include `GRANT`, `REVOKE`, and
`ALTER DEFAULT PRIVILEGES`.

Release note: None

mgartner committed:
sql: rename function privileges to routine privileges

This commit renames "function" privileges to "routine" privileges
because these privileges are also used for procedures, and "routine" is
the more general term that includes both functions and procedures.

Also, the `tree.FuncObj` type  has been renamed to `tree.RoutineObj` to
be more general. Related types, variables, and functions have be
similarly renamed.

Release note: None

mgartner committed:
sql: add basic support for procedure privileges

This commit implements the basic logic for granting, revoking, and
authorizing privileges on procedures. These operations perform the same
as privilege-related operations on functions, so the tests in
`procedure_privileges` have largely been copied from `udf_privileges`.
There's a number of statements that are not yet supported, so many of
the copied tests are commented out for now. They will be addressed in a
future commit.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/111133)
<!-- Reviewable:end -->

rharding6373 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice work!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111133#-:-Neyrr8G8fPxzxIHFCMz:bcmw7ss)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball, @mgartner, and @rafiss)

___
*[`pkg/sql/logictest/testdata/logic_test/show_default_privileges` line 7 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/111133#-Neynw_I7BgTCIRsFnlJ:-Neynw_I7BgTCIRsFnlK:b-vadkzi) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e0b21dd4a2f040026c2c2f1dfede7aa975c6675/pkg/sql/logictest/testdata/logic_test/show_default_privileges#L7)):*
> ```
> ----
> role  for_all_roles  object_type  grantee  privilege_type  is_grantable
> root  false          routines     public   EXECUTE         false
> ```

If we're going to expose "routines" as an object type, we should open a docs issue if we haven't already to clarify this term.  Searching for`object_type` in the docs leads to https://www.cockroachlabs.com/docs/stable/schema-design-update#alter-database-schema-objects and https://www.cockroachlabs.com/docs/stable/sql-grammar#target_object_type, for example. IIUC users should be using "functions" or "procedures" as the object_type in `ALTER` or `GRANT`. AFAICT postgres doesn't surface this kind of info, but I haven't dug very deep so this may be irrelevant from a compatibility perspective.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111133)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball, @rafiss, and @rharding6373)

___
*[`pkg/sql/logictest/testdata/logic_test/show_default_privileges` line 7 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/111133#-Neynw_I7BgTCIRsFnlJ:-Nf1hQlVAdAUSg2KJXdu:bojwp05) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e0b21dd4a2f040026c2c2f1dfede7aa975c6675/pkg/sql/logictest/testdata/logic_test/show_default_privileges#L7)):*
I don't see "functions" listed specifically in that first link, so I don't know what we'd change there. The second link should update automatically since that auto-generated from `sql.y`.

> AFAICT postgres doesn't surface this kind of info, but I haven't dug very deep so this may be irrelevant from a compatibility perspective.

Right - I kind just went with my gut on this, but happy to change the nomenclature for any of these commands if someone has a strong opinion. I'm curious what @cockroachdb/sql-foundations thinks.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Looks great!

> Reviewed 2 of 6 files at r1, 52 of 52 files at r4, 39 of 39 files at r5, 14 of 14 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111133#-:-NfFZxCq2Sqd1fmy-s_4:b-2t11ox)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @rafiss, @rharding6373, and @rhu713)



<!-- Sent from Reviewable.io -->

mgartner commented:
`TestGrantExecuteToPublicOnAllFunctions` is timing-out in CI, so I'll have to get to the bottom of that. I'm having trouble reproducing on my machine.

mgartner committed:
upgrades: skip TestGrantExecuteToPublicOnAllFunctions under race

TestGrantExecuteToPublicOnAllFunctions is very slow under the `--race`
test flag, often timing-out in CI. This commit skips this test under the
`--race` flag.

Release note: None

mgartner commented:
TFTRs!

bors r+

ecwall commented on pkg/sql/parser/testdata/alter_default_privileges:
```diff
@@ -62,6 +62,15 @@ ALTER DEFAULT PRIVILEGES GRANT ALL ON FUNCTIONS TO foo, bar -- fully parenthesiz
 ALTER DEFAULT PRIVILEGES GRANT ALL ON FUNCTIONS TO _, _ -- identifiers removed
 
+error
+ALTER DEFAULT PRIVILEGES GRANT ALL ON PROCEDURES TO foo
```
Is this being implemented in another PR? Or do PROCEDURES not have default privileges?

mgartner replied:
It's a syntax error in Postgres:

```
marcus=# ALTER DEFAULT PRIVILEGES GRANT ALL ON PROCEDURES TO foo;
ERROR:  42601: syntax error at or near "PROCEDURES"
LINE 1: ALTER DEFAULT PRIVILEGES GRANT ALL ON PROCEDURES TO foo;
```

And this decision is backed-up by the [Postgres docs](https://www.postgresql.org/docs/current/sql-alterdefaultprivileges.html):

> For this command, functions include aggregates and procedures. The words FUNCTIONS and ROUTINES are equivalent in this command. (ROUTINES is preferred going forward as the standard term for functions and procedures taken together. In earlier PostgreSQL releases, only the word FUNCTIONS was allowed. It is not possible to set default privileges for functions and procedures separately.)

ecwall replied:
Ah ok, good to know.

Do we normally return a syntax error for situations like this? It is not very user friendly IMO since it does not convey that "It is not possible to set default privileges for functions and procedures separately."

mgartner replied:
My aim was to mimic Postgres, but a more user-friendly error could be nice. Feel free to create an issue and we can put it on the backlog.
